lib/signs/utilities/early_am_suppression.ex

```diff
@@ -0,0 +1,302 @@
+          messages,
+          schedule,
+          status,
+          current_time
+        )
+    end
+  end
+
+  defp get_mezzanine_early_am_content(
+         messages,
+         sign,
+         schedule,
+         statuses,
+         current_time
+       ) do
+    {top_scheduled, bottom_scheduled} = schedule
+    {top_message, bottom_message} = messages
+    {top_status, bottom_status} = statuses
+
+    top_content =
+      if top_status == :none,
+        do: {top_message, %Message.Empty{}},
+        else:
+          get_early_am_content(
+            sign,
+            {top_message},
+            top_scheduled,
+            top_status,
+            current_time
+          )
+
+    bottom_content =
+      if(bottom_status == :none,
+        do: {bottom_message, %Message.Empty{}},
+        else:
+          get_early_am_content(
+            sign,
+            {bottom_message},
+            bottom_scheduled,
+            bottom_status,
+            current_time
+          )
+      )
+      |> case do
+        # If bottom line has status :none, then it could be a paging headway message
+        {%Headways.Paging{destination: destination, range: range}, _} ->
+          {%Headways.Top{destination: destination, vehicle_type: :train},
+           %Headways.Bottom{range: range}}
+
+        bottom_content ->
+          bottom_content
+      end
+
+    {top_content, bottom_content}
+  end
+
+  defp get_early_am_content(
+         sign,
+         messages,
+         {scheduled, destination},
+         status,
+         current_time
+       ) do
+    cond do
+      status == :fully_suppressed ->
+        {%EarlyAm.DestinationTrain{destination: destination},
+         %EarlyAm.ScheduledTime{
+           scheduled_time: scheduled
+         }}
+
+      status == :partially_suppressed ->
+        case filter_early_am_messages(messages, sign.id) do
+          [] ->
+            # If no valid predictions, try fetching headways
+            case Signs.Utilities.Headways.get_messages(sign, current_time) do
+              # If no headways are returned, default to timestamp message
+              {%Message.Empty{}, %Message.Empty{}} ->
+                {%EarlyAm.DestinationTrain{
+                   destination: destination
+                 },
+                 %EarlyAm.ScheduledTime{
+                   scheduled_time: scheduled
+                 }}
+
+              {headway_top, headway_bottom} ->
+                # Make sure routes is nil so that destination is used as headsign
+                {%{headway_top | destination: destination, routes: nil}, headway_bottom}
+            end
+
+          [message] ->
+            {message, %Content.Message.Empty{}}
+
+          messages ->
+            List.to_tuple(messages)
+        end
+    end
+  end
+
+  defp filter_early_am_messages(messages, sign_id) do
+    Tuple.to_list(messages)
+    |> Enum.reject(fn
+      message ->
+        match?(%Headways.Top{}, message) or match?(%Headways.Bottom{}, message) or
```

Comment:
I think the following may be an unhandled case, but I'm not fully confident I've followed the logic correctly:

If this is JFK, and the special paging logic triggers during normal message generation, we might get a `GenericPaging` with an embedded "half" of a prediction or headway message that needs suppressing. Even if we tried to match on it here, the other half of the message would be inside the other `GenericPaging`, which might not trigger suppression. If this is a valid case, I think we'd need to "unpack" the embedded messages before sending them to `get_early_am_content`, so we could handle both parts of each page at the same time.